torch/csrc/jit/mobile/debug_info.cpp

```diff
@@ -9,8 +11,78 @@
 namespace torch {
 namespace jit {
 
+namespace {
+
+// This function construct stacktrace with module hierarchy
+// Module hierarchy will contain information about where in the
+// module hierarchy this source is. For example if conv2d op
+// exist in hierarcy A->B->C->Conv2d with type annotations of
+// A -> TopM, B->MyModule, C->SomeModule, then module hierarchy
+// will be TopM(A).MyModule(B).SomeModule(C).Conv2d(conv)
+// Source level stack information will be from model source code.
+std::pair<std::string, std::string> getStackTraceWithModuleHierarchy(
+    const DebugInfoPair& source_callstack,
+    const std::string& root_scope_string,
+    const std::string& top_module_type_name) {
+  constexpr size_t kSourceRange = 1;
+  constexpr size_t kModuleInstanceInfo = 2;
+  std::vector<StackEntry> entries;
+
+  const SourceRange& range = source_callstack.first;
+  InlinedCallStackPtr callstack_ptr = source_callstack.second;
+  std::string module_info =
+      root_scope_string + "(" + top_module_type_name + ")";
+  if (!callstack_ptr) {
+    // If not cs then top level node
+    entries.emplace_back(StackEntry{"FunctionName_UNKNOWN", range});
+    std::ostringstream ss;
+    format_stack_trace(ss, entries);
+    std::string stack_trace = "Module hierarchy:" + module_info + "\n";
```

Comment:
nit: on both branches (here and line 73) you add the "Module hierarchy:" info line, why not:
1) factor them out - i.e. seems you only need handle 'if (callstack_ptr)' case first, handling the immediate callstack (line 37~42) is common logic for both cases;
2) push this string to the "std::ostringstream ss" before calling format_stack_trace, as it already takes the ostreamstream;

But this is just my personal style preference, you can safely ignore :)

Reply:
Good observation. Let me do that.